### PR TITLE
Preprocessing

### PR DESCRIPTION
**Pull request by neelsomani**
**Not ready to merge.**

`handler.py` is an AWS Lambda function to get the local bounds for chunks of data, and scale chunks of data to global bounds. `aggregate.py` reads in local bounds for all chunks of data (computed by `handler.py`) and aggregates them into the global bounds.

Lambda function is getting the error "Process exited before completing request," but runs fine locally.

### CODE REVIEW

**Comment by neelsomani** [comment]
**Ready for review.** Everything works now.

Here's how it scales w.r.t. the number of chunks of data:
1 object: 29.955233097076416 seconds
10 objects: 64.29478168487549 seconds
100 objects: 152.88286185264587 seconds

Next steps could be to merge this PR and generalize the interface in another, or to make more preprocessing scripts in this PR.

**Comment by neelsomani** [comment]
Hi @jcarreira, I've added a preprocessing module called `Preprocessing.py`, which is ready for review. An alternative that I was thinking about was maybe modeling the interface off of the sklearn preprocessing interface. Specifically, we would have one module called `preprocessing.py`, which would have the classes `MinMaxScaler`, `NormalScaler`, etc. within it. Then the user would initialize the scaler:

`s = preprocessing.MinMaxScaler(s3_src_bucket, s3_dest_bucket, 0.0, 1.0)`

and run it with something like `s.normalize()`.

**Comment by neelsomani** [comment]
Hey @jcarreira, updates:

* Added hashing trick + test for hashing trick
* Normalization should work when the memory limit on the lambda is increased

Ready for review, or we can add other preprocessing if needed.

File: python/frontend/cirrus/cirrus/NormalScaler.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    start_global = time.time()
+    print("LocalRange took {0} seconds...".format(start_global - start_bounds))
+
+    client = boto3.client("s3")
+    f_ranges = {}
+    for i in objects:
+        obj = client.get_object(Bucket=s3_bucket_input, Key=str(i) + "_bounds")["Body"].read()
+        d = json.loads(obj.decode("utf-8"))
+        for idx in d:
+            u_x_sq, u_x, n = d[idx]
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/NormalScaler.py]: These variable names are not readable.

File: python/frontend/cirrus/cirrus/lambdas/HashingTrickHandler.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    # Replace the appropriate columns for this row.
+    c = set([int(i) for i in columns])
+    for idx, r in enumerate(data):
+        row_map = {}
+        for col, val in r:
+            s_c = int(col)
+            if s_c in c:
+                h = hash(str(val)) % N
+                if h not in row_map:
+                    row_map[h] = 0
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/HashingTrickHandler.py]: I think this should be ```=1```.

File: python/frontend/cirrus/cirrus/lambdas/handler.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import MinMaxHandler
+import NormalHandler
+import HashingTrickHandler
+from serialization import *
+import time
+import random
+
+cluster = False
+
+def handler(event, context):
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/handler.py]: This function is way too big. 100 of lines and no empty lines.

File: python/frontend/cirrus/cirrus/lambdas/redis/__init__.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+from redis.client import Redis, StrictRedis
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/redis/__init__.py]: The redis code should not be committed.

File: python/frontend/cirrus/cirrus/lambdas/toml.py [hunk 5]
```diff
@@ -0,0 +1,1039 @@
+"""Python module which parses and emits TOML.
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/toml.py]: This should not be here.

File: python/frontend/cirrus/cirrus/serialization.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
+# Helper functions for S3.
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/serialization.py]: This file is a mix of s3 functions, LambdaThread class and serialization.
I would create a separate file for the LambdaThread. The rest you can put in a utils.py file.

**Comment by jcarreira** [comment]
You use python's ```hash()``` to hash the data. It seems [1] that for newer versions this function is non-deterministic which is not what we want. We should find how to make this deterministic.

https://stackoverflow.com/questions/27522626/hash-function-in-python-3-3-returns-different-results-between-sessions

File: python/frontend/cirrus/cirrus/HashingTrick.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import boto3
+from collections import deque
+from serialization import LambdaThread, get_all_keys
+from threading import Thread
+import random
+
+class HashingThread(LambdaThread):
+    def __init__(self, s3_bucket_input, s3_bucket_output, s3_key, columns, N, invocation):
+        Thread.__init__(self)
+        self.d = {
```
[the remark this answers is not in the document]
Comment (neelsomani) [reply python/frontend/cirrus/cirrus/HashingTrick.py]: I think that the dictionary makes the most sense here, given that it will be passed directly to the Lambda function, but feel free to correct me if I'm wrong here

File: python/frontend/cirrus/cirrus/HashingTrick.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            "s3_bucket_input": s3_bucket_input,
+            "s3_bucket_output": s3_bucket_output,
+            "s3_key": s3_key,
+            "action": "HASHING_TRICK",
+            "columns": columns,
+            "N": N,
+            "redis": "1",
```
[the remark this answers is not in the document]
Comment (neelsomani) [reply python/frontend/cirrus/cirrus/HashingTrick.py]: Fixed

File: python/frontend/cirrus/cirrus/utils.py [hunk 9]
```diff
@@ -1,4 +1,104 @@
...
+            final_objects.append(o)
+    print("Chunks after pruning: {0}".format(len(final_objects)))
+    return final_objects
+
+
+def get_data_from_s3(client, src_bucket, src_object, keep_label=False):
+    # Return a 2D list, where each element is a row of the dataset.
+    # print("Getting bytes from boto3")
+    b = client.get_object(Bucket=src_bucket, Key=src_object)["Body"].read()
+    # print("Got {0} bytes".format(len(b)))
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/utils.py]: Remove these commented prints.

File: python/frontend/cirrus/cirrus/LambdaThread.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+
+class LambdaThread(Thread):
+    def run(self):
+        l_client = boto3.client("lambda")
+        failure = 1
+        overall = time.time()
+        # Prevent lambdas from launching multiple times
+        self.lamdba_dict["dupe_nonce"] = (random.random() * 1000000) // 1.0
+        while failure < 4:
```
[the remark this answers is not in the document]
Comment (neelsomani) [reply python/frontend/cirrus/cirrus/LambdaThread.py]: Done, moved to utils

**Comment by neelsomani** [comment]
Hey @jcarreira, I just finished cleaning up the code, further refactoring, and running pylint - PTAL

File: python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+def hash_data(data, columns, N):
+    """ Replace the appropriate columns for this row. """
+    col_set = set([int(i) for i in columns])
+    for idx, row in enumerate(data):
+        row_map = {}
+        for col, val in row:
+            s_c = int(col)
+            if s_c in col_set:
+                hasher = hashlib.sha256()
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py]: This should not be created every time. Also we should use MurmurHash.

https://pypi.org/project/mmh3/

File: python/frontend/cirrus/cirrus/feature_hashing.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+            "action": "FEATURE_HASHING",
+            "columns": columns,
+            "N": N,
+            "use_redis": "1"
+        }
+
+
+def feature_hashing(s3_bucket_input, s3_bucket_output, columns, N, objects=()):
+    """ Take a list of integer values (column indices) to perform the feature hashing
+    with, for N buckets. """
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/feature_hashing.py]: This "with," doesn't make sense

File: python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+""" Helper functions for feature hashing """
+
+import hashlib
+import json
+import struct
+
+import boto3
+
+
+def hash_data(data, columns, N):
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py]: ```N``` is not a good name here. I would use hash_size.

File: python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    """ Replace the appropriate columns for this row. """
+    col_set = set([int(i) for i in columns])
+    for idx, row in enumerate(data):
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py]: I don't think this works if we hash a subset of the columns. In that case we will mix the non-hashed columns with the hashed columns.

File: python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py (hunk 11 above)
Comment (neelsomani) [reply python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py]: I think it should be created every time if we're using hashlib, but I've switched to MurmurHash regardless

File: python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py (hunk 14 above)
Comment (neelsomani) [reply python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py]: Hmm my interpretation was that those kind of collisions are just a part of how feature hashing works. How should I handle it differently? In `InputReader.cpp`, it looks like we're either hashing [all of the features] or excluding the [non-categorical features].

[all of the features]: https://github.com/jcarreira/cirrus/blob/master/src/InputReader.cpp#L892
[non-categorical features]: https://github.com/jcarreira/cirrus/blob/master/src/InputReader.cpp#L592

**Comment by neelsomani** [comment]
Hey @jcarreira, this latest push is actually all of the files after pylint. Any errors that are still in pylint either can't be fixed (e.g., it's saying too many parameters are being passed to a function that cannot be refactored, like the invocation of a LambdaThread), or it would break forwards compatibility (removing parentheses from print statements). Using pylint for Python2

File: python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py (hunk 14 above)
Comment (jcarreira) [reply python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py]: Yes, the C++ was done for a single case.

The way I would do this is to move all columns that are not hashed to the range ```[0, num_of_columns_not_hashed-1]```. All the other columns go to the range ```[num_of_columns_not_hashed, num_of_columns_not_hashed+N]```.

**Comment by jcarreira** [comment]
Please fix conflicts.

File: python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py (hunk 14 above)
Comment (neelsomani) [reply python/frontend/cirrus/cirrus/lambdas/feature_hashing_helper.py]: Since I'm not sure how many columns there are in the data without traversing through it, I opted to shift the columns not hashed to the end of the range instead (`[0, num_columns_hashed]` and the original columns are shifted by `num_columns_hashed`)

**Comment by neelsomani** [comment]
Hey @jcarreira, fixed the merge conflicts

File: python/frontend/cirrus/cirrus/lambdas/deploy.sh [hunk 15]
```diff
@@ -0,0 +1,5 @@
+rm bundle.zip
+zip -9r bundle.zip handler.py mmh3* min_max_helper.py ../utils.py normal_helper.py redis/ rediscluster/ toml.py redis.toml feature_hashing_helper.py
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/deploy.sh]: These folders redis/, rediscluster/, etc.. need to be setup in this script.

File: python/frontend/cirrus/tests/preprocessing_test.py [hunk 16]
```diff
@@ -0,0 +1,282 @@
+""" Tests for the preprocessing module for Cirrus """
```
Comment (jcarreira) [inline python/frontend/cirrus/tests/preprocessing_test.py]: This test doesn't seem runnable. I would create a main that runs all the tests here.

Also add instructions on how to run this. E.g, where is the data used for the test.

File: python/frontend/cirrus/cirrus/lambdas/deploy.sh (hunk 15 above)
Comment (neelsomani) [reply python/frontend/cirrus/cirrus/lambdas/deploy.sh]: Created a separate script called `setup.sh`

File: python/frontend/cirrus/cirrus/lambdas/handler.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+        timer.timestamp()
+
+
+def normal_scaling_handler(s3_client, data, labels, event):
+    """ Scale to a unit normal range """
+    if event["action"] == "LOCAL_RANGE":
+        print("Getting local data ranges...")
+        bounds = normal_helper.get_data_ranges(data)
+        print("Putting ranges in S3...")
+        min_max_helper.put_bounds_in_db(
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/lambdas/handler.py]: I may be wrong but I don't think this code will work. put_bounds_in_db calls get_keys_values which assumes the input data has a "max" key. normal_helper.get_data_ranges(data) doesn't seem to comply with this.

File: python/frontend/cirrus/tests/preprocessing_test.py [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+        objects = get_all_keys(s3_bucket_input)
+    timer.timestamp()
+    if preprocess:
+        timer.set_step("Running preprocessing")
+        Preprocessing.normalize(s3_bucket_input, s3_bucket_output,
+                                Normalization.MIN_MAX, min_v, max_v,
+                                objects, True, False, skip_bounds)
+        timer.timestamp()
+    timer.set_step("Running threads for each object")
+    launch_lambdas(SimpleTest, objects, 400, s3_bucket_input,
```
Comment (jcarreira) [inline python/frontend/cirrus/tests/preprocessing_test.py]: Any reason why this code launches lambdas? Doesn't seem like a good idea.

File: python/frontend/cirrus/cirrus/preprocessing.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    def load_libsvm(path, s3_bucket):
+        """ Load a libsvm file into S3 in the specified bucket. """
+        client = boto3.client("s3")
+        timer = Timer("LOAD_LIBSVM").set_step("Reading file")
+        data = sklearn.datasets.load_svmlight_file(path)[0]
+        timer.timestamp().set_step("Starting loop")
+        batch = []
+        batch_num = 1
+        batch_size = 0
+        for row in range(data.shape[0]):
```
Comment (jcarreira) [inline python/frontend/cirrus/cirrus/preprocessing.py]: This code is very slow.

My code below seems to be 8x faster. Can you see if it works here:

```
from enum import Enum
import sklearn.datasets
from context import cirrus

import boto3
import sys 
from cirrus.utils import serialize_data, Timer

path = "/home/ubuntu/joao/cirrus-neel/cirrus/python/frontend/cirrus/tests/criteo.small.svm"

timer = Timer("LOAD_LIBSVM").set_step("Reading file")
data = sklearn.datasets.load_svmlight_file(path)[0]
timer.timestamp()

print(data.shape)

batch = [0] * 200000 # this line is important
batch_num = 1 
batch_size = 0 

timer.timestamp().set_step("To lil")
lil = data.tolil(copy=False) # convert to list of lists format
timer.timestamp().set_step("Traverse lil")


for i, (row, data) in enumerate(zip(lil.rows, lil.data)):
  curr_row = []
  for j, val in zip(row, data):
    curr_row.append((j, val))
  batch[batch_size] = curr_row
  batch_size += 1
timer.timestamp()
```

File: python/frontend/cirrus/tests/preprocessing_test.py [hunk 20]
```diff
@@ -0,0 +1,375 @@
+""" Tests for the preprocessing module for Cirrus.
+
+You can get the test data from
```
Comment (jcarreira) [inline python/frontend/cirrus/tests/preprocessing_test.py]: We need a test for normalization
